src/qrexec-agent/qrexec-agent.c

```diff
@@ -70,6 +70,95 @@ void DumpRequestList(void)
 }
 #endif
 
+struct _connection_info
+{
+    HANDLE handle;
+    int connect_domain;
+    int connect_port;
+};
+
+#define MAX_FDS 256
+static struct _connection_info connection_info[MAX_FDS];
+
+static void register_vchan_connection(HANDLE handle, int domain, int port)
+{
+    for (int i = 0; i < MAX_FDS; i++)
+    {
+        if (connection_info[i].handle == NULL)
+        {
+            EnterCriticalSection(&g_DaemonCriticalSection);
+            connection_info[i].handle = handle;
+            connection_info[i].connect_domain = domain;
+            connection_info[i].connect_port = port;
+            LeaveCriticalSection(&g_DaemonCriticalSection);
+            return;
+        }
+    }
+    LogError("No free slot for child %d (connection to %d:%d)", (int)handle, domain, port);
+}
+
+static void release_connection(int id)
+{
+    struct msg_header hdr;
+    struct exec_params params;
+
+    HANDLE handle = connection_info[id].handle;
+    hdr.type = MSG_CONNECTION_TERMINATED;
+    hdr.len = sizeof(struct exec_params);
+    params.connect_domain = connection_info[id].connect_domain;
+    params.connect_port = connection_info[id].connect_port;
+    if (libvchan_send(g_DaemonVchan, &hdr, sizeof(hdr)) != sizeof(hdr))
+    {
+        LogError("send (MSG_CONNECTION_TERMINATED hdr)");
+        return;
+    }
+    if (libvchan_send(g_DaemonVchan, &params, sizeof(params)) != sizeof(params))
+    {
+        LogError("send (MSG_CONNECTION_TERMINATED data)");
+        return;
+    }
+    CloseHandle(handle);
+    connection_info[id].handle = NULL;
+    return;
+}
+
+static BOOL handle_is_running(HANDLE hProcess)
+{
+    if (hProcess == INVALID_HANDLE_VALUE)
+    {
+        return FALSE;
+    }
+    DWORD dwRetval = WaitForSingleObject(hProcess, 0);
+
+    switch(dwRetval)
+    {
+    case WAIT_OBJECT_0:
+        return FALSE;
+    case WAIT_TIMEOUT:
+        return TRUE;
+    default:
+        return FALSE;
+    }
+    return FALSE;
+}
+
+static void check_connections()
+{
+    for (int i = 0; i < MAX_FDS; i++)
+    {
+        HANDLE handle = connection_info[i].handle;
+        if (handle != NULL)
+        {
+            if (!handle_is_running(handle))
+            {
+                EnterCriticalSection(&g_DaemonCriticalSection);
```

Comment:
At the very least reading `connection_info[i].handle` should be already under the lock - otherwise another thread could release the slot in the meantime (resulting in releasing the same slot twice, at best). But it's safer to put the whole loop under the lock already.